contents/docs/libraries/next-js/_snippets/app-router-setup.mdx

```diff
@@ -0,0 +1,237 @@
+import posthog from 'posthog-js'
+import { PostHogProvider } from 'posthog-js/react'
+import { useEffect } from 'react'
+
+export function PHProvider({
+  children,
+}: {
+  children: React.ReactNode
+}) {
+    useEffect(() => {
+      posthog.init(process.env.NEXT_PUBLIC_POSTHOG_KEY, {
+        api_host: process.env.NEXT_PUBLIC_POSTHOG_HOST,
+        person_profiles: 'identified_only',
+        capture_pageview: false // Disable automatic pageview capture, as we capture manually
+      })
+  }, []);
+
+  return <PostHogProvider client={posthog}>{children}</PostHogProvider>
+}
+```
+
+</MultiLanguage>
+
+PostHog's `$pageview` autocapture relies on page load events. Since Next.js acts as a single-page app, this event doesn't trigger on navigation and we need to capture `$pageview` events manually. 
+
+To do this, we set up a `PostHogPageView` component to listen to URL path changes:
+
+<MultiLanguage>
+
+```js
+// app/PostHogPageView.jsx
+'use client'
+
+import { usePathname, useSearchParams } from "next/navigation";
+import { useEffect } from "react";
+import { usePostHog } from 'posthog-js/react';
+
+export default function PostHogPageView() {
+  const pathname = usePathname();
+  const searchParams = useSearchParams();
+  const posthog = usePostHog();
+  useEffect(() => {
+    // Track pageviews
+    if (pathname && posthog) {
+      let url = window.origin + pathname
+      if (searchParams.toString()) {
+        url = url + `?${searchParams.toString()}`
+      }
+      posthog.capture(
+        '$pageview',
+        {
+          '$current_url': url,
+        }
+      )
+    }
+  }, [pathname, searchParams, posthog])
+  
+  return null
+}
+```
+
+```ts
+// app/PostHogPageView.tsx
+'use client'
+
+import { usePathname, useSearchParams } from "next/navigation";
+import { useEffect } from "react";
+import { usePostHog } from 'posthog-js/react';
+
+export default function PostHogPageView() : null {
+  const pathname = usePathname();
+  const searchParams = useSearchParams();
+  const posthog = usePostHog();
+  useEffect(() => {
+    // Track pageviews
+    if (pathname && posthog) {
+      let url = window.origin + pathname
+      if (searchParams.toString()) {
+        url = url + `?${searchParams.toString()}`
+      }
+      posthog.capture(
+        '$pageview',
+        {
+          '$current_url': url,
+        }
+      )
+    }
+  }, [pathname, searchParams, posthog])
+  
+  return null
+}
+```
+
+</MultiLanguage>
+
+Then, import the `PHProvider` component into your `app/layout` file and wrap your app with it. We also dynamically import the `PostHogPageView` component and include it as a child of `PHProvider`.
+
+> **Why is `PostHogPageView` dynamically imported?** It contains the [`useSearchParams`](https://nextjs.org/docs/app/api-reference/functions/use-search-params) hook, which [deopts](https://nextjs.org/docs/messages/deopted-into-client-rendering) the entire app into client-side rendering if it is not dynamically imported.
```

Comment:
I might be missing something from elsewhere, but my original confusion was more about `PHProvider` than `PostHogPageView`. Since PHProvider is a client component, and is now wrapping my entire app, hasn't my whole app just de-opted into client-side rendering anyway?

This could well be something I'm just getting wrong - but I would have thought that having a client component so close to the root of the app will prevent the magical component streaming from the server. But maybe nextjs is smart enough to start streaming server component children? If so it would be great if there was a blog post about that to link to or something? (Obviously Posthog doesn't need to be responsible for nextjs documentation, but it *feels* like a big deal to wrap the entire app in a client component)

Reply:
Here's the relevant FAQ entry for this:

https://github.com/PostHog/posthog.com/blob/87e229835030ce52f4c4c77624576a04ef79745b/contents/docs/libraries/next-js/index.mdx?plain=1#L124-L129

Which links to details here: https://react.dev/reference/rsc/use-client#why-is-copyright-a-server-component